include/lo2s/perf/reader.hpp

```diff
@@ -0,0 +1,368 @@
+    {
+    public:
+        ParseError(const std::string& what) : std::runtime_error(what)
+        {
+        }
+
+        ParseError(const std::string& what, int error_code);
+    };
+
+    TracepointEvent() : id_(-1)
+    {
+    }
+    TracepointEvent(bool enable_on_exec, int event_id);
+    TracepointEvent(bool enable_on_exec, const std::string& name);
+
+    friend bool operator==(const TracepointEvent& lhs, const TracepointEvent& rhs)
+    {
+        return lhs.id_ == rhs.id_;
+    }
+
+    friend bool operator<(const TracepointEvent& lhs, const TracepointEvent& rhs)
+    {
+        return lhs.id_ < rhs.id_;
+    }
+
+    friend bool operator>(const TracepointEvent& lhs, const TracepointEvent& rhs)
+    {
+        return lhs.id_ > rhs.id_;
+    }
+
+    void parse_format();
+
+    void as_syscall()
+    {
+        attr_.sample_type |= PERF_SAMPLE_IDENTIFIER;
+    }
+
+    const auto& fields() const
+    {
+        return fields_;
+    }
+
+    const auto& common_fields() const
+    {
+        return common_fields_;
+    }
+
+    const auto& field(const std::string& name) const
+    {
+        for (const auto& field : fields())
+        {
+            if (field.name() == name)
+            {
+                return field;
+            }
+        }
+        throw std::out_of_range("field not found");
+    }
+
+    int id()
+    {
+        return id_;
+    }
+
+    std::string name() const
+    {
+        return name_;
+    }
+
+    static std::vector<std::string> get_tracepoint_event_names();
+
+private:
+    void parse_format_line(const std::string& line);
+
+    const static std::filesystem::path base_path_;
+    int id_;
+    std::string name_;
+    std::vector<tracepoint::EventField> common_fields_;
+    std::vector<tracepoint::EventField> fields_;
+};
+
+/**
+ * Contains a raw event for testing
+ */
+class RawEvent : public PerfEvent
+{
+public:
+    RawEvent(bool enable_on_exec);
+};
+
+/**
+ * Contains an opened instance of PerfEvent.
+ * Use PerfEvent.open() to construct an object
+ */
+class PerfEventInstance
+{
+public:
+    PerfEventInstance();
+    PerfEventInstance(PerfEvent& ev, std::variant<Cpu, Thread> location, int group_fd,
+                      int cgroup_fd);
+
+    PerfEventInstance(PerfEventInstance&) = delete;
+    PerfEventInstance& operator=(const PerfEventInstance&) = delete;
+
+    PerfEventInstance(PerfEventInstance&& other)
+    {
+        std::swap(fd_, other.fd_);
+        std::swap(ev_, other.ev_);
+    }
+
+    PerfEventInstance& operator=(PerfEventInstance&& other)
+    {
+        std::swap(fd_, other.fd_);
+        std::swap(ev_, other.ev_);
+        return *this;
+    }
+
+    void enable();
+    void disable();
+
+    void set_output(const PerfEventInstance& other_ev);
+    void set_syscall_filter();
+
+    int get_fd() const
+    {
+        return fd_;
+    }
+
+    bool is_valid() const
+    {
+        return fd_ >= 0;
+    };
+
+    template <class T>
+    T read()
+    {
+        T val;
```

Comment:
`static_assert(std::is_pod_v<T> == true);` will fail, as UserspaceReadFormat has a user-defined constructor.

Reply:
Remove the constructor. If I see it correctly, the only thing it does is to set the members to 0. This can be ensured by different ways (memset) if there are places where this might cause issues